fw/ss_skb.c

```diff
@@ -1008,6 +1008,94 @@ ss_skb_chop_head_tail(struct sk_buff *skb_head, struct sk_buff *skb,
 	return 0;
 }
 
+/**
+ * Chop @head and @tail bytes at head and end of the message (skb circular
+ * list), iterating over the list if nessessary.
+ * @return 0 on success and -(error code) on error.
+ *
+ * Message length should be greater than the sum of @head and @trail,
+ * otherwise an error is retuned.
+ */
+int
+ss_skb_list_chop_head_tail(struct sk_buff **skb_list_head, 
+			   size_t head, size_t trail)
+{
+
+	struct sk_buff *skb, *skb_hd;
+	size_t sum;
+	int R;
+	skb = *skb_list_head;
+	if (likely(skb->next == skb))
+		goto single_buff;
+	if (likely(head))
+		goto head_and_probably_trail;
+	if (likely(trail))
+		goto trail_only;
+	/* nothing to chop */
+	return 0;
+
+single_buff:
+	/* There is the only 1 buffer in the list */
+	sum = head + trail;
+	if (WARN_ON(skb->len <= sum))
+		return -EBADMSG;
+	if (sum == 0)
+		/* Nothing to chop */
+		/* This check is mostly for jumps from branches below */
+		return 0;
+	return ss_skb_chop_head_tail(skb, skb, head, trail);
```

Comment:
Calling `ss_skb_chop_head_tail()` only once looks like a good optimization: we really can expect that in most the cases we'll have only one skb here.